pilot/pkg/serviceregistry/serviceentry/servicediscovery.go

```diff
@@ -677,3 +681,42 @@ func selectorChanged(old, curr model.Config) bool {
 	n := curr.Spec.(*networking.ServiceEntry)
 	return !reflect.DeepEqual(o.WorkloadSelector, n.WorkloadSelector)
 }
+
+// automatically allocates IPs for service entry services without an address field
+// if the hostname is not a wildcard, or when resolution is not NONE. The IPs are allocated
+// from the 240.240.0.0 subnet that is not reachable outside the pod. When DNS capture is
+// enabled, Envoy will resolve the DNS to these IPs. The listeners for TCP services will also
+// be set up on these IPs.
+func autoAllocateIPs(services []*model.Service) []*model.Service {
+	// i is everything from 240.240.0.(j) to 240.240.255.(j)
+	// j is everything from 240.240.(i).1 to 240.240.(i).254
+	// we can capture this in one integer variable.
+	// given X, we can compute i by X/255, and j is X%255
+	// To avoid allocating 240.240.(i).255, if X % 255 is 0, increment X.
+	// For example, when X=510, the resulting IP would be 240.240.2.0 (invalid)
+	// So we bump X to 511, so that the resulting IP is 240.240.2.1
+	maxIPs := 255 * 255 // are we going to exceeed this limit by processing 64K services?
+	x := 0
+	for _, svc := range services {
+		// we can allocate IPs only if
+		// 1. the service has resolution set to static/dns. We cannot allocate
+		//   for NONE because we will not know the original DST IP that the application requested.
+		// 2. the address is not set (0.0.0.0)
+		// 3. the hostname is not a wildcard
+		if svc.Address == constants.UnspecifiedIP && !svc.Hostname.IsWildCarded() &&
+			svc.Resolution != model.Passthrough {
+			x++
+			if x%255 == 0 {
+				x++
+			}
+			if x >= maxIPs {
+				log.Errorf("out of IPs to allocate for service entries")
+				return services
+			}
+			thirdOctet := x / 255
+			fourthOctet := x % 255
+			svc.AutoAllocatedAddress = fmt.Sprintf("240.240.%d.%d", thirdOctet, fourthOctet)
```

Comment:
This should not be user configurable because we do not want this to be ever reachable outside the pod. Its automatically allocated, meant for resolution inside the pod by Envoy, and never meant to be used outside of the pod. This address range is a Class E range that is reserved/unallocated (https://en.wikipedia.org/wiki/Reserved_IP_addresses). Its the same range we use for multicluster replicated control planes, and documented in our docs as well.

Reply:
Thanks for the link, wasn't aware of that.